### PR TITLE
feat(hardhat): use hardhat-deploy for deployments

### PR DESCRIPTION
**Pull request by jflint256**
Add `hardhat-deploy` plugin and use it for deployments.
Make sure `Spender` gets deployed from `deployerSpender (accounts[0])` and `Router` from `deployerRouter (accounts[1])`
Tested on sepolia.

### CODE REVIEW

File: packages/hardhat/deploy/001-deploy-Spender.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import { network } from 'hardhat';
+import verify from '../utils/verify/verify';
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const { deployments, getNamedAccounts } = hre;
+  const { deploy, log } = deployments;
+
+  const { deployerSpender } = await getNamedAccounts();
+
+  const args: any[] = [];
```
Comment (abrkn) [inline packages/hardhat/deploy/001-deploy-Spender.ts:13]: If possible use `unknown[]`

File: packages/hardhat/deploy/001-deploy-Spender.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+  const spender = await deploy('Spender', {
+    from: deployerSpender,
+    args,
+    log: true,
+    autoMine: true,
+  });
+
+  if (!devChains.includes(network.name)) {
+    log('------- verifying started --------');
```
Comment (abrkn) [inline packages/hardhat/deploy/001-deploy-Spender.ts:23]: `log('Verifying...');`

Once we start with the `----- verifying started ----` the next is `!!!!! verifying failed !!!!` and we never win the competition over attention

File: packages/hardhat/deploy/002-deploy-SifiV1Router01.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import { network } from 'hardhat';
+import verify from '../utils/verify/verify';
+
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  const { deployments, getNamedAccounts } = hre;
+  const { deploy, get, log } = deployments;
+
+  const { deployerRouter } = await getNamedAccounts();
+
+  const SpenderAddress = (await get('Spender')).address;
```
Comment (abrkn) [inline packages/hardhat/deploy/002-deploy-SifiV1Router01.ts:13]: `const spenderAddress = `

File: packages/hardhat/helper-hardhat-config.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+export const devChains = ['hardhat', 'localhost', 'sepolia', 'goerli'];
```
Comment (abrkn) [inline packages/hardhat/helper-hardhat-config.ts:1]: Consider `DEV_CHAINS`, but up to you. Gets tiresome reading caps

File: packages/hardhat/utils/verify/verify.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable node/no-unpublished-import */
+import { run } from "hardhat";
+
+const verify = async (contactAdress: string, args: any[]) => {
+  console.log("verifying contract...");
```
Comment (abrkn) [inline packages/hardhat/utils/verify/verify.ts:5]: Where are these double quotes coming from? Do we have a broken `.prettierrc`?

File: packages/hardhat/utils/verify/verify.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  try {
+    await run("verify:verify", {
+      address: contactAdress,
+      constructorArguments: args,
+    });
+  } catch (e: any) {
+    if (e.message.toLowerCase().includes("already verified")) {
```
Comment (abrkn) [inline packages/hardhat/utils/verify/verify.ts:12]: - `catch (error: any)`
- Double quotes again
- `console.error(error.message);`
- Consider `process.exit(1);` on error

**Review by abrkn — CHANGES_REQUESTED** [review]

File: packages/hardhat/deploy/001-deploy-Spender.ts (hunk 1 above)
Comment (xykota) [inline packages/hardhat/deploy/001-deploy-Spender.ts:13]: Use `const args: unknown[]` if possible

File: packages/hardhat/deploy/001-deploy-Spender.ts (hunk 2 above)
Comment (xykota) [inline packages/hardhat/deploy/001-deploy-Spender.ts:23]: `log('Verifying...');`

File: packages/hardhat/deploy/002-deploy-SifiV1Router01.ts (hunk 3 above)
Comment (xykota) [inline packages/hardhat/deploy/002-deploy-SifiV1Router01.ts:13]: `const spenderAddress = `

File: packages/hardhat/deploy/002-deploy-SifiV1Router01.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+  const sifiV1Router01 = await deploy('SifiV1Router01', {
+    from: deployerRouter,
+    args,
+    log: true,
+    autoMine: true,
+  });
+
+  if (!devChains.includes(network.name)) {
+    log('------- verifying started --------');
```
Comment (xykota) [inline packages/hardhat/deploy/002-deploy-SifiV1Router01.ts:30]: Same as above

File: packages/hardhat/helper-hardhat-config.ts (hunk 4 above)
Comment (xykota) [inline packages/hardhat/helper-hardhat-config.ts:1]: Consider `DEV_CHAINS`, up to you

File: packages/hardhat/utils/verify/verify.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+/* eslint-disable node/no-unpublished-import */
+import { run } from "hardhat";
+
+const verify = async (contactAdress: string, args: any[]) => {
+  console.log("verifying contract...");
+  try {
+    await run("verify:verify", {
```
Comment (xykota) [inline packages/hardhat/utils/verify/verify.ts:7]: Where are double quotes coming from? Is prettier working in the repo?

File: packages/hardhat/utils/verify/verify.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+      address: contactAdress,
+      constructorArguments: args,
+    });
+  } catch (e: any) {
```
Comment (xykota) [inline packages/hardhat/utils/verify/verify.ts:11]: - `catch (error: any) {`
- console.error(error.message);`
- Consider `process.exit(1);` on error

**Review by xykota — CHANGES_REQUESTED** [review]

**Pull request merged**